File: impacted-test-engine/src/test/java/com/teamscale/test_impacted/engine/executor/SimpleTestDescriptor.java
package com.teamscale.test_impacted.engine.executor;

import org.junit.platform.engine.TestDescriptor;
import org.junit.platform.engine.TestExecutionResult;
import org.junit.platform.engine.UniqueId;
import org.junit.platform.engine.support.descriptor.AbstractTestDescriptor;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static org.junit.platform.engine.TestExecutionResult.successful;

/** A basic implementation of a {@link TestDescriptor} that can be used during tests. */
public class SimpleTestDescriptor extends AbstractTestDescriptor {

	private final Type type;

	private boolean shouldSkip = false;

	private final List<TestDescriptor> dynamicTests = new ArrayList<>();

	private TestExecutionResult executionResult = successful();

	private SimpleTestDescriptor(UniqueId uniqueId, Type type, String displayName) {
		super(uniqueId, displayName);
		this.type = type;
	}

	@Override
	public Type getType() {
		return type;
	}

	/** Marks the test as being skipped. */
	public SimpleTestDescriptor skip() {
		this.shouldSkip = true;
		return this;
	}

	/** Whether the test should be skipped. */
	public boolean shouldBeSkipped() {
		return shouldSkip;
	}

	/**
	 * The dynamic child tests that should be simulated or just an empty list if the test does not contain dynamic
	 * tests.
	 */
	public List<TestDescriptor> getDynamicTests() {
		return dynamicTests;
	}

	/** Sets the execution result that the engine should report when simulating the test's execution. */
	public SimpleTestDescriptor result(TestExecutionResult executionResult) {
		this.executionResult = executionResult;
		return this;
	}

	public TestExecutionResult getExecutionResult() {
		return this.executionResult;
	}

	/** Creates a {@link TestDescriptor} for a concrete test case without children. */
	public static SimpleTestDescriptor testCase(UniqueId uniqueId) {
		return new SimpleTestDescriptor(uniqueId, Type.TEST, getSimpleDisplayName(uniqueId));
	}

	private static String getSimpleDisplayName(UniqueId uniqueId) {
		return uniqueId.getSegments().get(uniqueId.getSegments().size() - 1).getValue();
	}

	/** Creates a {@link TestDescriptor} for a dynamic test case which registers children during test execution. */
	public static TestDescriptor dynamicTestCase(UniqueId uniqueId, TestDescriptor... dynamicTestCases) {
		SimpleTestDescriptor simpleTestDescriptor = new SimpleTestDescriptor(uniqueId, Type.CONTAINER_AND_TEST,
				getSimpleDisplayName(uniqueId));
		simpleTestDescriptor.dynamicTests.addAll(Arrays.asList(dynamicTestCases));
		return simpleTestDescriptor;
	}

	/**
	 * Creates a {@link TestDescriptor} for a test container (e.g. a test class or test engine) containing other
	 * {@link TestDescriptor} children.
	 */
	public static SimpleTestDescriptor testContainer(UniqueId uniqueId, TestDescriptor... children) {
		SimpleTestDescriptor result = new SimpleTestDescriptor(uniqueId, Type.CONTAINER,
				getSimpleDisplayName(uniqueId));
		for (TestDescriptor child : children) {
			result.addChild(child);
		}
		return result;
	}
}

File: impacted-test-engine/src/test/kotlin/com/teamscale/test_impacted/engine/executor/SimpleTestDescriptor.kt
package com.teamscale.test_impacted.engine.executor

import org.junit.platform.engine.TestDescriptor
import org.junit.platform.engine.TestExecutionResult
import org.junit.platform.engine.UniqueId
import org.junit.platform.engine.support.descriptor.AbstractTestDescriptor

/**
 * Represents a simple implementation of a test descriptor for managing test cases and test containers within a
 * test engine framework. This class is designed to provide functionality for defining the structure and behavior
 * of test descriptors.
 *
 * @constructor Private constructor for creating an instance of this class. Instances can only be created through
 * companion object functions.
 *
 * @param uniqueId The unique identifier for the test descriptor.
 * @param type The type of the test descriptor, represented by [TestDescriptor.Type].
 * @param displayName The display name for the test descriptor.
 */
class SimpleTestDescriptor private constructor(
	uniqueId: UniqueId,
	private val type: TestDescriptor.Type,
	displayName: String
) : AbstractTestDescriptor(uniqueId, displayName) {
	private var shouldSkip = false

	/**
	 * A mutable list that holds dynamic test descriptors for a test case. These tests are registered dynamically
	 * during test execution and can include child tests or nested dynamic executions.
	 */
	val dynamicTests = mutableListOf<TestDescriptor>()

	/**
	 * Represents the execution result of a test, indicating whether the test was successful, failed, or aborted.
	 * This property is updated during the test execution process when the result of the test is determined.
	 */
	var executionResult: TestExecutionResult = TestExecutionResult.successful()
		private set

	override fun getType() = type

	/** Marks the test as being skipped.  */
	fun skip(): SimpleTestDescriptor {
		this.shouldSkip = true
		return this
	}

	/** Whether the test should be skipped.  */
	fun shouldBeSkipped() = shouldSkip

	/** Sets the execution result that the engine should report when simulating the test's execution.  */
	fun result(executionResult: TestExecutionResult): SimpleTestDescriptor {
		this.executionResult = executionResult
		return this
	}

	companion object {
		/** Creates a [TestDescriptor] for a concrete test case without children.  */
		fun testCase(uniqueId: UniqueId) =
			SimpleTestDescriptor(uniqueId, TestDescriptor.Type.TEST, getSimpleDisplayName(uniqueId))

		private fun getSimpleDisplayName(uniqueId: UniqueId) =
			uniqueId.segments[uniqueId.segments.size - 1].value

		/** Creates a [TestDescriptor] for a dynamic test case which registers children during test execution.  */
		fun dynamicTestCase(uniqueId: UniqueId, vararg dynamicTestCases: TestDescriptor): TestDescriptor {
			val simpleTestDescriptor = SimpleTestDescriptor(
				uniqueId, TestDescriptor.Type.CONTAINER_AND_TEST,
				getSimpleDisplayName(uniqueId)
			)
			simpleTestDescriptor.dynamicTests.addAll(listOf(*dynamicTestCases))
			return simpleTestDescriptor
		}

		/**
		 * Creates a [TestDescriptor] for a test container (e.g., a test class or test engine) containing other
		 * [TestDescriptor] children.
		 */
		fun testContainer(uniqueId: UniqueId, vararg children: TestDescriptor): SimpleTestDescriptor {
			val result = SimpleTestDescriptor(
				uniqueId, TestDescriptor.Type.CONTAINER,
				getSimpleDisplayName(uniqueId)
			)
			children.forEach { child ->
				result.addChild(child)
			}
			return result
		}
	}
}
